metasv/extract_pairs.py

```diff
@@ -69,40 +69,48 @@ def discordant_with_normal_orientation(aln, mate, chr_tid, chr_start, chr_end, i
+    mate = None
+    for bam_handle in bam_handles:
+        try:
+            mate = bam_handle.mate(aln)
+        except ValueError:
+            pass
+        if mate is not None:
+            return mate
+    return mate
+
+
+def extract_read_pairs(bam_handles, region, prefix, extract_fns, pad=0, max_read_pairs = EXTRACTION_MAX_READ_PAIRS,
                        truncation_pad_read_extract = EXTRACTION_TRUNCATION_PAD,  
                        max_interval_len_truncation = EXTRACTION_MAX_INTERVAL_TRUNCATION, sv_type=''):
     logger = logging.getLogger("%s-%s" % (extract_read_pairs.__name__, multiprocessing.current_process()))
 
     extract_fn_names = [extract_fn.__name__ for extract_fn in extract_fns]
-    logger.info("Extracting reads from %s for region %s with padding %d using functions %s" % (
-        bamname, region, pad, extract_fn_names))
+    logger.info("Extracting reads for region %s with padding %d using functions %s" % (
+        region, pad, extract_fn_names))
 
     chr_name = str(region.split(':')[0])
     chr_start = int(region.split(':')[1].split("-")[0]) - pad
     chr_end = int(region.split(':')[1].split('-')[1]) + pad
 
     selected_pair_counts = [0] * len(extract_fn_names)
-    aln_list = []
-
     start_time = time.time()
 
-    bam = pysam.Samfile(bamname, "rb")
-    chr_tid = bam.gettid(chr_name)
     if chr_start < 0:
+        regions_to_extract = []
         logger.error("Skipping read extraction since interval too close to chromosome beginning")
     else:
         # Read alignments from the interval in memory and build a dictionary to get mate instead of calling bammate.mate() function
+        regions_to_extract = [(chr_name, chr_start, chr_end)]
         if abs(chr_end-chr_start)>max_interval_len_truncation and sv_type in ["INV","DEL","DUP"]:
             # For large SVs, middle sequences has no effect on genotyping. So, we only extract reads around breakpoints to speed up
-            truncate_start = chr_start + pad + truncation_pad_read_extract 
+            truncate_start = chr_start + pad + truncation_pad_read_extract
             truncate_end = chr_end -  (pad + truncation_pad_read_extract)
             logger.info("Truncate the reads in [%d-%d] for %s_%d_%d" % (truncate_start,truncate_end,chr_name,chr_start,chr_end))
-            aln_list = [aln for aln in bam.fetch(chr_name, start=chr_start, end=truncate_start-1) if not aln.is_secondary] + \
-                       [aln for aln in bam.fetch(chr_name, start=truncate_end+1, end=chr_end) if not aln.is_secondary]
-        else:        
-            aln_list = [aln for aln in bam.fetch(chr_name, start=chr_start, end=chr_end) if not aln.is_secondary]
+            regions_to_extract = [(chr_name, chr_start, truncate_start-1), (chr_name, truncate_end+1, chr_end)]
+
+    aln_list = [aln for (chr_, start_, end_) in regions_to_extract for bam_handle in bam_handles for aln in bam_handle.fetch(chr_, start=start_, end=end_) if not aln.is_secondary]
 
```

Comment:
This is just a small step towards more general read extraction. For a complex SV involving multiple breakpoints (can be > 2), we need to extract reads around each breakpoint. Currently, we only assemble from at most 2 breakpoints, but in the future, it will need to be more than 2.